server/src/main/java/org/candlepin/controller/ContentManager.java

```diff
@@ -306,28 +303,56 @@ public Content updateContent(ContentDTO update, Owner owner, boolean regenerateE
         updated = this.contentCurator.create(updated);
 
         this.ownerContentCurator.updateOwnerContentReferences(owner,
-            Collections.<String, String>singletonMap(entity.getUuid(), updated.getUuid()));
+            Collections.singletonMap(entity.getUuid(), updated.getUuid()));
 
         // Impl note:
         // This block is a consequence of products and contents not being strongly related.
         log.debug("Updating {} affected products", affectedProducts.size());
-        ContentDTO cdto = this.modelTranslator.translate(updated, ContentDTO.class);
+        ContentDTO updatedContent = this.modelTranslator.translate(updated, ContentDTO.class);
 
         // TODO: Should we bulk this up like we do in importContent? Probably.
         for (Product product : affectedProducts) {
             log.debug("Updating affected product: {}", product);
+
             ProductDTO pdto = this.modelTranslator.translate(product, ProductDTO.class);
+            addContent(pdto, updatedContent);
 
-            ProductContentDTO pcdto = pdto.getProductContent(cdto.getId());
-            if (pcdto != null) {
-                pdto.addContent(cdto, pcdto.isEnabled());
+            // Impl note: This should also take care of our entitlement cert regeneration
+            this.productManager.updateProduct(pdto, owner, regenerateEntitlementCerts);
+        }
 
-                // Impl note: This should also take care of our entitlement cert regeneration
-                this.productManager.updateProduct(pdto, owner, regenerateEntitlementCerts);
+        return updated;
+    }
+
+    private ProductContentDTO findContent(Set<ProductContentDTO> contents, String id) {
+        if (id == null) {
+            return null;
+        }
+        for (ProductContentDTO content : contents) {
+            if (id.equals(content.getContent().getId())) {
+                return content;
             }
         }
+        return null;
+    }
 
-        return updated;
+    private void addContent(ProductDTO product, ContentDTO update) {
+        if (update == null || update.getId() == null) {
+            throw new IllegalArgumentException("update references incomplete content");
+        }
+
+        Set<ProductContentDTO> productContent = product.getProductContent();
```

Comment:
That is a problem with our current use of dtos. DTO is breaking encapsulation by definition. Using them as a makeshift domain objects is only making it worse. We are modifying the all over the place which we should not do.

Reply:
We've already run into problems a number of times before by doing this. Making assumptions about how the DTOs map their data isn't a good practice here. As soon as you do, you're coupling the internal logic of two completely separate things -- one of which we have very little control over now.

Should ContentManager be using DTOs these days? Absolutely not. It's an area that's very deserving of a refactor. But that's not reason enough to exacerbate the problem.

Reply:
We are apparently agreeing about the problem. What do you suggest w should do? I refactored the code so that the collection are at least not modified in place, but coupling is still there.

Reply:
In the interest of getting this PR merged, and not adding additional work, I think the solution
for now is good. But we should find a better way of doing this in another task.

I spent sometime trying to understand why we even use DTOs in this way in the ContentManager.
Firstly, using DTOs in Product/ContentManager is no problem in itself, if what we're doing is
simply accepting a DTO as an update from a resource endpoint and only call getters on it.
And I agree with both of you, we shouldn't be altering the data on the DTO.
The reason (which seems uniquely affecting ContentManager but not ProductManager) looks to me to be:

What we want to do:
- We're fetching entities from the DB (Content and Product) and want to process them (involves altering the data on them).
- Then we want to pass on that data to the ProductManager (because they're intrinsically connected), BUT, the ProductManager
  accepts DTOs as input, so we have to translate the entities into DTOs and then pass them to ProductManager.

What we're actually doing:
1. Fetching entities from the DB (alternate versions of content, affected products etc.)
2. We're explicitly using a translator and turning them into DTOs (_ONLY because we want to later send them to ProductManager, which accepts DTOs_)
3. We're processing/altering them (here is the current problem).
4. Sending them off to ProductManager.

What I think we should be doing:
1. Fetching entities from DB.
2. We're processing/altering them. **
3. We're translating them to DTOs.
4. Sending them to ProductManager.

** If we have hibernate issues in step 2 here, as some comments in the code imply,
where altering collections on entities is a problem, then we have some alternatives:
a. Up until now we used DTOs (but now this is a problem, so something else should be used)
b. We can clone the entities (so that they are not hibernate-managed, and alter freely)
c. We can create new data containers (you could call them DTOs) that provide the encapsulation
we need.

I think a is not an option, and c is overkill (because we'd still need to re-translate into api DTOs), and prefer option b.

Reply:
@Januson 
We should be avoiding directly modifying the collection and assuming it will update the DTO's state. That last part is what makes this class dependent on the inner workings of the DTO. We should, instead, create a new collection of the content we want, and assign that to the DTO.

@nikosmoum 
Some history on why DTOs were ever used in the Product/Content manager:
Originally, it only accepted model objects, as the expectation was that it would take a modified product, fetch a new instance from the DB exactly how it's represented on the backend, and do all its calculations based on the modified local (potentially attached) instance, and the detached instance it fetched representing the current DB state.

However, due to how we've configured Hibernate (primarily, auto-commit/flush), there are a number of operations which cause Hibernate to be "helpful" and persist some changes to managed objects, even if we haven't explicitly called .merge or .update. The first version had a number of weird bugs where collections on products and content would not be reflected in version comparisons, because Hibernate had persisted the change for us as soon as we did *any* select on the DB against the table owning the object -- even if it was for a different entity/instance.

This led to a number of choices, and after weighing the pros/cons of them, we, perhaps erroneously, opted to use DTOs at this layer to allow us to pass in the changes without risking putting a managed object in an unknown state.

Since then, we've made a number of other changes, such as making product and content immutable, which may fix the issue entirely by means of telling Hibernate to never change the object; we've added the service adapter model, which is how most of these changes are made today; and I'm currently working on a refresh refactor which pulls much of the work out of these managers.

It's entirely probable that in the very near future, we'll have a clear view of what these managers need to do, and how we should go about using them. The resource refactor should be the final nail in the coffin of passing the DTOs into the managers, and will likely be the optimal point for addressing the remaining components of the managers in question.

Reply:
@Ceiu @Januson Alright, we can address the long term impact after those changes. For the short-term changes in this PR then, the change should be: iterate the DTO collection items and add them to an entirely new collection, modify the new collection as needed, then add it back to the DTO, right?